tokenlists/typing.py

```diff
@@ -1,92 +1,133 @@
-from copy import deepcopy
-from datetime import datetime as DateTime
-from typing import Dict, Iterator, List, Optional, Union
+from datetime import datetime
+from typing import Dict, List, Optional
 
-import dataclassy as dc
+from pydantic import AnyUrl
+from pydantic import BaseModel as _BaseModel
+from pydantic import validator
 from semantic_version import Version  # type: ignore
 
 ChainId = int
 
 TagId = str
-URI = str
 
 TokenAddress = str
 TokenName = str
 TokenDecimals = int
 TokenSymbol = str
 
 
-@dc.dataclass(frozen=True, slots=True)
-class TokenInfo:
+class BaseModel(_BaseModel):
+    def dict(self, *args, **kwargs):
+        if "exclude_unset" not in kwargs:
+            kwargs["exclude_unset"] = True
+
+        return super().dict(*args, **kwargs)
+
+    class Config:
+        froze = True
+
+
+class TokenInfo(BaseModel):
     chainId: ChainId
     address: TokenAddress
     name: TokenName
     decimals: TokenDecimals
     symbol: TokenSymbol
-    logoURI: Optional[str] = None
+    logoURI: Optional[AnyUrl] = None
     tags: Optional[List[TagId]] = None
-    extensions: Optional[Dict[str, Union[str, int, bool]]] = None
-
-    @classmethod
-    def from_dict(cls, data: Dict) -> "TokenInfo":
-        data = deepcopy(data)
-        return cls(**data)  # type: ignore
-
-    def to_dict(self) -> Dict:
-        data = dc.asdict(self)
-        if self.logoURI is None:
-            del data["logoURI"]
-        if self.tags is None:
-            del data["tags"]
-        if self.extensions is None:
-            del data["extensions"]
-        return data
+    extensions: Optional[dict] = None
+
+    @validator("address")
+    def address_must_hex(cls, v: str):
+        if not v.startswith("0x") or set(v) > set("x0123456789abcdefABCDEF") or len(v) % 2 != 0:
+            raise ValueError("Address is not hex")
+
+        address_bytes = bytes.fromhex(v[2:])  # NOTE: Skip `0x`
+
+        if len(address_bytes) != 20:
+            raise ValueError("Address is incorrect length")
+
+        return v
+
+    @validator("decimals")
+    def decimals_must_be_uint8(cls, v: TokenDecimals):
+        if not (0 <= v < 256):
+            raise ValueError(f"Invalid token decimals: {v}")
+
+        return v
+
+    @validator("extensions")
+    def extensions_must_contain_simple_types(cls, d: Optional[dict]) -> Optional[dict]:
+        if not d:
+            return d
 
+        # `extensions` is `Dict[str, Union[str, int, bool, None]]`, but pydantic mutates entries
+        for val in d.values():
+            if not isinstance(val, (str, int, bool)) and val is not None:
+                raise ValueError("")
```

Comment:
Is there supposed to be an error message here?

Reply:
ah yes good catch lol
```suggestion
                raise ValueError("Incorrect extension field value")
```